### ISSUE TITLE
Make QuickFolders compatible with Thunderbird 91. (ESR 2021/22)

### ISSUE THREAD

**Issue by RealRaven2000** [RealRaven2000/QuickFolders#114]
There are a number of things that don't work in the latest beta for Thunderbird 86, so far I have discovered:

- Options screen is misaligned (especially support and layout tabs) due to missing grid support
- Search popup have stopped working
- (add more later)

**Comment by RealRaven2000** [comment]
Quick fixes for the first two items listed above:

[QuickFolders-wx-5.3pre54.zip](https://github.com/RealRaven2000/QuickFolders/files/6015632/QuickFolders-wx-5.3pre54.zip)

To install, download the zip file and drag into Thunderbird Add-ons manager. This version should also be compatible with Thunderbird 78, so feel free to test there as well and leave comments below:

**Comment by aprojanos71** [comment]
Works fine!
Thx!

(You are very quick!) 👍🏼

**Comment by RealRaven2000** [comment]
> Works fine!
> Thx!
> 
> (You are very quick!) 👍🏼

Yeah, I already had found that grid bug earlier ans just discovered the search not working while messing around with quickFilters. That's why I decided to create a catch-all bug for Thunderbird 86 like in quickFilters, so I can hit the ground running later this year when the ESR rolls around once more. The changes aren't quite as fundamental as having to drop the legacy flag altogether (as Thunderbird 78 required) but my guess is they will be more numerous this year because Thunderbird will force developers to publish new Add-ons as "Mail extensions" - with native APIs only. 

Which means they can do more internal changes / refactoring and care about internal dependencies of our code less - if we still use the legacy methods of "hacking the internals" - like all my Add-ons we have to once again blame ourselves if we can't keep up. At the moment I am solving the problems with "wrapper code"  - so I am writing functions that "translate" between both behaviors, but I had dropped a lot of that in Tb78 because I was forced to fork my code into a new branch - which effectivelyu doubles my workload whenever I try to backport bug fixes / features for Thunderbird 68. Once I release for 88 I will probably freeze all work on Thunderbird 68 because it will only have diminishing returns at that stage.

**Comment by RealRaven2000** [comment]
First two items fixed in 5.3 - released 22/02/2021. Please add more defects / regressions for Tb86 / later below.

**Comment by RealRaven2000** [comment]
More defects found in Tb 88:

- subfolders are not shown in subfolder dropdowns that are displayed when right-clicking a Tab
- folder navigation next / previous "sibling" folder don't work.

These are caused by folder.subFolders having been changed from an nsIMutableArray to an Array. I added some wrapper code so that it will work both in Tb78 and in the newer versions as well. quickFix tested in Thunderbird 87.0b1

[QuickFolders-wx-5.4pre16.zip](https://github.com/RealRaven2000/QuickFolders/files/6072493/QuickFolders-wx-5.4pre16.zip)

**Comment by RealRaven2000** [comment]
Next problem in Tb87 

- opening documentation / content URLS / websites in a tab fails. THis is due to a change in contentTabs, which now requires a uri attribute to be passed instead of a contentPage attribute.

The patched version simply passes both so it remains backwards compatible with Tb78, too:
[QuickFolders-wx-5.4pre18.zip](https://github.com/RealRaven2000/QuickFolders/files/6072744/QuickFolders-wx-5.4pre18.zip)

**Comment by RealRaven2000** [comment]
Next problem in Tb87

- advanced tab-specific properties was malformed because of lacking grid support:
![image](https://user-images.githubusercontent.com/6314674/109801940-14ccf300-7c17-11eb-9013-f7001b010509.png)
these elements should be aligned vertically.

Fixed version:
[Uploading QuickFolders-wx-5.4pre23.zip…]()

**Comment by RealRaven2000** [comment]
I published Release 5.4 on ATN today

- Fixed: Support Pages did not open anymore as content Tabs have a different parameter set.
- Fixed: layout of advanced tab properties broken due to deprecated grid element.
- Fixed: Copy / Move folders uses new MailServices.copy object

**Comment by RealRaven2000** [comment]
> Super!
> But what is ATN?

an initialism for [addons.thunderbird.net](https://addons.thunderbird.net/en-US/thunderbird/addon/quickfolders-tabbed-folders/) - where I release all my updates that are automatically deployed to my users. It used to be AMO (addons.mozilla.org).

**Comment by RealRaven2000** [comment]
Here is a test version for Thunderbird 88:
[QuickFolders-wx-5.4.2pre23.zip](https://github.com/RealRaven2000/QuickFolders/files/6217389/QuickFolders-wx-5.4.2pre23.zip)

I hope addons.thunderbird.net will support uploading something with max-version number 8.0 so that all betas can be accepted.

**Comment by RealRaven2000** [comment]
To do: create list of changes for @jobisoft to document what's chanbging in Thunderbird ESR 2021/22

e.g. util.moveMessages() function

```
         let isListArray = util.versionGreaterOrEqual(util.ApplicationVersion, "85"); 
        if (isListArray) 
          messageList.push(Message);
        else
          messageList.appendElement(Message , false);
```

**Comment by RealRaven2000** [comment]
To Do: revisit `QuickFolders.Util.allFoldersIterator `- it may be possible to simplify this in a manner that is backwards compatible with Thundebird 78. at the moment it creates a nsIMutableArray, which may already be unnecessary within Thunderbird 78!

**Comment by jobisoft** [comment]
The reported changes have been added to DTN (I missed ftvItem).

Thanks!

**Comment by RealRaven2000** [comment]
Open items: 

- we still cannot open any "about:config" mini-windows to set stuff like "debug options" (by right-clicking the Debug Mode checkbox). Obviously the config editor was changed to run in a content tab
- Theme support needs to be improved, see #186 (affects Tb78 and Tb91 in differing ways)

**Comment by RealRaven2000** [comment]
Latest test version for Thunderbird 91 (tested in Tb daily 92.0a1):

[QuickFolders-mx-5.7pre64.zip](https://github.com/RealRaven2000/QuickFolders/files/6869710/QuickFolders-mx-5.7pre64.zip)

To install this preview, download the zip file and drag it into Thunderbird Add-ons manager.

**Comment by RealRaven2000** [comment]
- Caused by deprecation of "fixIterator" the quickMove / quickJump functions stopped working. Here is a fixer version using normal Arrays tested in  Thunderbird daily  92.0a1 64bit.

[QuickFolders-mx-5.7pre106.zip](https://github.com/RealRaven2000/QuickFolders/files/6902790/QuickFolders-mx-5.7pre106.zip)

**Comment by RealRaven2000** [comment]
- Fixed the missing (and badly styled) [Buy License] button at the bottom of the options dialog. This was already a problem since Tb78. Apparently it is part of a "shadow DOM" which ignores the style sheets of the main dialog.
- showing the debug settings (and other advanced settings) via a small window that filters "about:config" stopped working in Tb91. To test , right-click any of the quickMove / reading list / recent folders button checkboxes or the debug mode checkbox in the options window:

![image](https://user-images.githubusercontent.com/6314674/127929306-ef929bd8-6924-44f9-bb4b-fada8338ffff.png)


[QuickFolders-mx-5.7pre120.zip](https://github.com/RealRaven2000/QuickFolders/files/6919846/QuickFolders-mx-5.7pre120.zip)

To install this preview, download the zip file and drag it into Thunderbird Add-ons manager.

**Comment by RealRaven2000** [comment]
Looks like `MailServices.copy.CopyMessages` was renamed to `MailServices.copy.copyMessages` : sigh.

Fresh preview with quick fix:

[QuickFolders-mx-5.7pre174.zip](https://github.com/RealRaven2000/QuickFolders/files/6947955/QuickFolders-mx-5.7pre174.zip)

To install this preview, download the zip file and drag it into Thunderbird Add-ons manager. Caution: contains fresh code from issue #187 - so if you haven't got any license you're a bit of a guinea pig right now. Just ironing out the kinks...

**Comment by RealRaven2000** [comment]
Fixed:  dragging the envelope from current folder toolbar (Navigation bar). The envelope icon represents the  currently selectged emails and dragging it failed in Thunderbird 91

![image](https://user-images.githubusercontent.com/6314674/128613212-9604da7c-76b5-4124-ae59-04f2671009ab.png)

Test Version below:

[QuickFolders-mx-5.7pre212.zip](https://github.com/RealRaven2000/QuickFolders/files/6949762/QuickFolders-mx-5.7pre212.zip)

**Comment by RealRaven2000** [comment]
Another drag + drop issue I found was: #190 - this involved being able to drag the mail icon on the Current Folder toolbar (center screen) in order to make for a very quick and easy drag operation to targets such as the "recent folders" toolbar button from a constant remembered screen location. The bug is addressed in this preview version:

[QuickFolders-mx-5.7pre212.zip](https://github.com/RealRaven2000/QuickFolders/files/6951224/QuickFolders-mx-5.7pre212.zip)
To install this preview, download the zip file and drag it into Thunderbird Add-ons manager.

**Comment by RealRaven2000** [comment]
Implemented in 5.7 - released on 10/08/2021
